New common pin handling

holger-dengler opened the pull request:
This PR introduce a common pin handling. All other implementations of pin entry has been removed.

The pin handling for user- and so-pins will now first use the values of environment variables and, if they are not set, prompt for input. The prompt is now also able to read from redirected stdin, which makes scripting a lot easier. Expect scripts should no longer necessary.

The "downside" is, that, for the first time, the env handling is working correctly at all. In some of the test-scripts, which rely on the previous (broken) behaviour, which may led to hick-ups, especially while most of the test-scritps use the env variables PKCS11_USER_PIN and PKCS11_SO_PIN for their own purpose. This may interfere with the new functionality here an there. The last commit in this PR contains a quickfix for such a situation in the expect-script for initializing the tokens. Hopefully this is enough for making Travis happy...

The PR also covers the missing documentation of the env variables.

Open discussion point:
The current implementation supports the env variables OCK_SO_PIN, OCK_USER_PIN and also PKCS11_USER_PIN. We should decide, if we switch completely to either PKCS11_* or OCK_* for the variable names.

holger-dengler commented:
Force push with cleanups, mainly to fix the Travis failures.

ifranzki commented on usr/lib/common/pin_get.h:
```diff
@@ -0,0 +1,86 @@
+struct pinbuf {
+    char *p;
+    size_t len;
+};
```
Is there a good reason why you introduce this struct? Why not simply use a zero terminated string?
Its essentially a buffer containing a string with a specific length. 

The only good reason I can imagine is to allow pins to include the \0 character in its payload.
But then you can't use strlen, strncmp, etc to operate on the pin contents (which you currently do in your implementation). 

In my opinion this makes it unnecessarily complicated. Just use a zero terminated string instead, and trust that the string is in fact zero terminated.

holger-dengler replied:
The problem for the pin_get_*() functions is, that we have 2-3 different source for the pin pointers, argv, env variable or the buffer from the input (getline()). Only the latter on is dynamic and only for this purpose, we need a char pointer and a size. Note: len is the length of allocated space in p, not the length of the input.

So, from a user point of view, the pin should be accessed only via the return value of the pin_get_*() or pin_prompt() functions. The stuct pinbuf variables has the only purpose, to keep track of the dynamically allocated memory. We do not really need a struct here, we could also just use a pointer to a char and a pointer to a size_t variable. But having a struct here, makes it a bit less complicated in my opinion.

BTW: I do not agree, that we should support \0 in pins. All strings, which are returned by pin_get_*() function, are null-terminated.

ifranzki replied:
> The problem for the pin_get_*() functions is, that we have 2-3 different source for the pin pointers, argv, env variable or the buffer from the input (getline()). Only the latter on is dynamic and only for this purpose, we need a char pointer and a size.

Yes, I understood that.  

> We do not really need a struct here, we could also just use a pointer to a char and a pointer to a size_t variable.

Do we really need the (allocated) size? Anywhere where we need the length, we can do `strlen()`. 
But yes, we do need the allocated pointer, e.g. to free it afterwards.

So instead of passing `struct pinbuf *buf` we could simply pass `char **buf` (or even `void **buf` if you like) and treat `buf` the same way as before, i.e. pass it to `pinbuf_free()`, `pin_get_user()`, etc.

holger-dengler replied:
But if we remove the length information, we can not wipe out the pin on free. And we loose the possibility to work with pre-allocated buffers (because without a length, we do not know, how much space has been allocated). 

What about the following:
Make `struct pinbuf`` only visible to pin_get.c and use a `typedef pinbuf_t struct pinbuf` in the header file. So all users need a pinbuf_t variable to work with the pin_get functions. I used such a mechanism for the buffer.c as well.

holger-dengler replied:
Hm... why not usig p11_buffer_t for that purpose? It has all what we need. Let's give it a try.

ifranzki replied:
> But if we remove the length information, we can not wipe out the pin on free.

Sure you can: just use `strlen()` to get the length to wipe. Wiping the terminating zero by another zero does not add much :-) 

> And we loose the possibility to work with pre-allocated buffers

Are you actually using this somewhere in the code? If not, I guess we don't need that possibility.

> Make struct pinbuf only visible to pin_get.c and use a typedef pinbuf_t struct pinbuf in the header file.

Thats better, but I still think you only need a `char*` and not a struct.

> Hm... why not using p11_buffer_t for that purpose? It has all what we need. Let's give it a try.

Again, I don't think all this is really needed. Lets please keep it as simple as possible.

holger-dengler replied:
ok

ifranzki commented:
Regarding PKCS11_USER_PIN vs OCK_USER_PIN:

The p11sak tool (in contrast to the other OCK tools) is supposed to be usable with other PKCS11 implementations as well (you can set env var PKCSLIB to point to a foreign PKCS11 implementation - this env var is also not mentioned in the man page....grrr) . It should in general not be OCK specific. The name PKCS11_USER_PIN is also more general than OCK_USER_PIN.

So at least for the p11sak tool I would actually prefer PKCS11_USER_PIN. For the other (OCK specific) tools, OCK_USER_PIN seems to be better....

Your approach to accept both certainly works, but this will actually mean that we have to keep both environment variables for a longer time... Removing it at a later time breaks an API, that's always bad to do.....

So I guess I would rather keep PKCS11_USER_PIN and not introduce yet another variable.

ifranzki commented on usr/lib/common/pin_get.c:
```diff
@@ -0,0 +1,136 @@
+    /* strip CR/LF (if any) */
+    (buf->p)[strcspn(buf->p, "\r\n")] = '\0';
+
+    buf->len = strlen(buf->p);
```
If the pin would contain a \n somewhere in the middle (not sure if getline would allow that at all), then you would get the wrong length here.
So I again wonder why you need the pinbuf struct at all, if the length is set via strlen anyway..

holger-dengler replied:
Yes and no. The pin should not contain \0, but the length calculation is anyway wrong. Len holds the allocated length, not the length of the pin. I'll fix that.

ifranzki commented on usr/lib/common/pin_get.c:
```diff
@@ -0,0 +1,136 @@
+
+    if ((buf->len && buf2.len) &&
+        (buf->len == buf2.len) &&
+        (!strncmp(buf->p, buf2.p, buf->len)))
```
if you are about to allow \0 within the pin, then you would need to use memcmp here.

holger-dengler replied:
The piece of code is wrong. the len variables do not represent the allocated space in p, not the length of the strings. I'll fix this in the next version.

ifranzki commented on usr/sbin/pkcsconf/pkcsconf.c:
```diff
@@ -105,47 +100,13 @@ int main(int argc, char *argv[])
-                }
-                memcpy(sopin, optarg, strlen(optarg) + 1);
-            }
+            pin_so = optarg;
```
OK its nit picking, but you loose the ability to detect that the -S flag is specified multiple times.  Maybe you should check if pin_so != NULL and issue the error message if so.

Same for -U and -n below.

holger-dengler replied:
ok, I'll fix this.

ifranzki commented on usr/lib/common/pin_get.c:
```diff
@@ -0,0 +1,136 @@
+        return;
+
+    if (buf->p)
+        memset(buf->p, 0, buf->len);
```
Use OPENSSL_cleanse here, memset may get optimized out by the compiler.

holger-dengler replied:
I would prefer not to introduce a dependecy to libcrypto here, just for cleaning the memory. Yes, the current implementation can be removed by compiler optimizations. But if you look into the OPENSSL_cleanse implementation, it is relatively easy to prevent that. 

pin_get is meant to be used in all ock tools. Should they all have a depency to libcrypto, just to wipe out the pin buffer? I would perfer to prevent the compiler optimization for this single line instead.

ifranzki replied:
we already link libcrypto everywhere (as well as many other libraries...).
But if you find a way to hinder the compiler to not optimize that out, fine with me.

ifranzki commented on usr/sbin/pkcsep11_session/pkcsep11_session.c:
```diff
@@ -90,120 +90,61 @@ CK_VERSION lib_version;
-    fflush(stdout);
-    ret = get_pin(&vhsmpin, &vhsmpinlen);
-    if (ret != 0) {
+    vhsmpin = pin_prompt(&buf_vhsm, "Enter the new VHSM PIN: ");
```
I guess we should be using pin_get_new() here.... (although it was not that way before, I know). We are asking the user to set a new pin, so maybe its better to require a confirmation of the pin entered here.

ifranzki replied:
Will need an update in testcases/init_vhsm.exp.in as well to provide the pin again to the 2nd prompt.

holger-dengler replied:
ok, I will have a look at that. Maybe I'll come back to you for the details.

holger-dengler replied:
This will not be in the next version, but I keep it on my to-do list.

ifranzki commented on man/man1/p11sak.1.in:
```diff
@@ -513,6 +516,16 @@ A custom path for p11sak_defined_attrs.conf can be set with the environment vari
+.SS "PKCS11_USER_PIN (deprecated)"
+Same as OCK_USER_PIN. If set, OCK_USER_PIN is prefered over PKCS11_USER_PIN.
+This environment variable is deprecated.
+.PP
```
Please also add the PKCSLIB  variable to specify the PKCS 11 implementation while you are at it.

holger-dengler replied:
Ok. But I will do it in separate commits.

ifranzki commented on testcases/init_token.sh.in:
```diff
@@ -61,17 +61,17 @@ expect {
 }
 expect {
-    "Re-enter the new user PIN: " { sleep .1; send $env(PKCS11_USER_PIN); send "\r"; }
+    "Re-enter new USER PIN: " { sleep .1; send $env(PKCS11_USER_PIN); send "\r"; }
```
You need to do the same in testcases/cleanup_vhsm.exp.in, testcases/init_token.sh.in, testcases/init_token.sh.in and testcases/init_vhsm.exp.in

holger-dengler replied:
Ok, I'll check this and change it accordingly.

holger-dengler replied:
Dito, not in the next version, but still on my to-do list.

ifranzki commented on man/man1/pkcsconf.1.in:
```diff
@@ -35,16 +35,36 @@ set the SO PIN
+is used or user will be prompted
 .IP "\fB-n\fP \fINEWPIN\fP" 10
 the new pin (for use when changing either the user pin or the SO pin; -u, -p
 and -P options); if not specified, user will be prompted
```
Maybe explicitly state that the new pin is never obtained from an environment variable.

holger-dengler replied:
Ok

jschmidb commented on man/man1/pkcsep11_session.1.in:
```diff
@@ -55,6 +55,16 @@ displayed or logged out.
+.SH "ENVIRONMENT VARIABLES"
+.SS "OCK_USER_PIN"
+The token PIN can be specified via the environment variable OCK_USER_PIN.
+If nothing is set, pkcsicsf will prompt for the token PIN interactively.
+.PP
```
pkcsicsf -> pkcsep11

jschmidb replied:
And also commit msg: pkcssep11_session -> pkcsep11_session

holger-dengler replied:
ok, will fix it.

holger-dengler replied:
This paragraph in the man page will be removed completely, because of the env variable rename.
The commit msg will be fixed.

jschmidb commented on man/man1/pkcsep11_migrate.1.in:
```diff
@@ -44,6 +44,16 @@ notation.
+If nothing is set, pkcsep11_migrate will prompt for the token PIN interactively.
+.PP
+.SS "PKCS11_USER_PIN (deprecated)"
+Same as OCK_USER_PIN. If set, OCK_USER_PIN is prefered over PKCS11_USER_PIN.
+This environment variable is deprecated.
```
prefered -> preferred

holger-dengler replied:
ok

ifranzki commented:
We had some discussions about the following:

- We all agree that passing pins via command line argument is unsecure, since it will then appear in the command history, top output, etc. Nevertheless we should keep this option.
- Passing pins via environment variables is better, but has the downside that a system wide set variable may be used by the executable without the user knowing that it is used. So it was suggested to add a program option to force it to prompt for the pin, even if the environment variable is set. 
- It was also suggested to an an option t read the pin from a file. That way the file can be access restricted by standard file system means. 
- It was suggested to have a verbose output of the tools showing where the pin was taken from, i.e. argument, env-var, file, prompt, etc. This should be printed when the tool runs in verbose/debug mode only.

Please check if/how these points could be incorporated into the common pin handling.

I guess the a force-for-promot option is a must-have, otherwise we make the tools more insecure than they were before.

I also guess that this was the reason why p121sak checked the environment variable only after an empty pin was entered. That way the user had a control when the env-var was used. However, that approach is not really usable for automation anyway.

jschmidb commented on man/man1/pkcsicsf.1.in:
```diff
@@ -68,6 +68,21 @@ the opencryptoki config file containing token configuration information
+interactively.
+.PP
+.SS "PKCS11_USER_PIN (deprecated)"
+Same as OCK_USER_PIN. If set, OCK_USER_PIN is prefered over PKCS11_USER_PIN.
+This environment variable is deprecated.
```
preferred

holger-dengler replied:
ok. Maybe I should grep for "prefered" and fix it all... <grml>

jschmidb commented on man/man1/pkcscca.1.in:
```diff
@@ -57,6 +57,21 @@ The PKCS slot number.
+interactively.
+.PP
+.SS "PKCS11_USER_PIN (deprecated)"
+Same as OCK_USER_PIN. If set, OCK_USER_PIN is prefered over PKCS11_USER_PIN.
+This environment variable is deprecated.
```
preferred

jschmidb commented on man/man1/pkcstok_migrate.1.in:
```diff
@@ -60,6 +60,24 @@ specifies the verbose level: \fInone\fP, error, warn, info, devel, debug
+is not specified, pkcstok_migrate will prompt for the token PIN interactively.
+.PP
+.SS "PKCS11_USER_PIN (deprecated)"
+Same as OCK_USER_PIN. If set, OCK_USER_PIN is prefered over PKCS11_USER_PIN.
+This environment variable is deprecated.
```
preferred

jschmidb commented on usr/lib/common/pkcs_utils.c:
```diff
@@ -86,7 +86,7 @@ void pkcs_trace(pkcs_trace_level_t level, const char *file, int line,
 }
 
-int compute_hash(int hash_type, int buf_size, char *buf, char *digest)
+int compute_hash(int hash_type, int buf_size, const char *buf, char *digest)
 {
```
Commit msg above: qualifiy -> qualify

holger-dengler replied:
Fixed.

ifranzki reviewed (changes requested):
We also had discussions about the environment variable naming. There are pros and cons for both names.

For backward compatibility we need to keep `PKCS11_USER_PIN` where it was used before.
So we need to keep `PKCS11_USER_PIN` in any case. 
Having 2 variable names that have basically the same meaning seems strange for me and just makes it more complicated.

So I would like to keep `PKCS11_USER_PIN` as environment variable name (and consequently also `PKCS11_SO_PIN`) and do not want to introduce new environment variables like `OCK_USER_PIN`.

At the end it is the PKCS11 pin that you specify with it, and thus `PKCS11_USER_PIN` fits, regardless for which executable you use it. The `OCK_` prefix should be used for environment variables hat are strictly specific to OpenCryptoki.

holger-dengler commented:
I agree, the current naming of the env variables are confusing. Lets continue with PKCS11_USER_PIN and PKCS11_SO_PIN for the pin_get functions. Maybe we can use OCK_USER_PIN and OCK_SO_PIN for our test scripts, just to prevent interference like in the expect scripts...

BTW: Which backward compatibility? The env variables did not worked in any of the tools as expected. They are only mentioned in the external documentation...

ifranzki commented:
> Maybe we can use OCK_USER_PIN and OCK_SO_PIN for our test scripts, just to prevent interference like in the expect scripts...

All tests under testcases do expect `PKCS11_USER_PIN` to be set. This has been that way for a long time, and I would like to keep it that way.

holger-dengler commented:
To your previous question about the pin source: the current implementation is able to read the the pin also from a pipe  or redirected stdin, which is also a very secure way for the pin input. You may use testcases/unit/pintest for some tests.

The following examples are all non-interactive.

```
$ echo "12345678" > pin.txt
$ testcases/unit/pintest -u < pin.txt
Enter the USER PIN:
pin: 12345678
$ echo "12345678" | testcases/unit/pintest -u
Enter the USER PIN:
pin: 12345678
$ printf "1234\n1234\n" | testcases/unit/pintest -n
Enter new PIN:
Re-enter new PIN:
pin: 1234
```

holger-dengler commented:
Yes, all tests had used PKCS11_USER_PIN and PKCS11_SO_PIN for a long time and use it today. But with this PR, the tools also care about these both variables. As you can see in the init-token case: pkcsconf is called and the expect-script tries to input the value from PKCS11_USER_PIN. But now, pkcsconf will not ask for a pin at all, because it is now aware of the env variable and read it directly from there...

But anyhow: I will rename the env variables to PKCS11_USER_PIN and PKCS11_SO_PIN. If this causes some trouble with the test-scripts, I'll fix the test scripts as well.

ifranzki commented:
The more I think about that the more I come to the conclusion that the whole environment stuff in the pin handling code causes more trouble than it solves. Adding a force-pin-prompt option to all the tools is a nightmare.

Why not drop all the environment code from the pin handling code, and have the user use `echo $ENVVAR | tool` or `tool --pin $ENVVAR` if one wants to have the pin in an environment variable. No extra code needed for that.

The only exception would be p11sak. It is documented that it uses the PKCS11_USER_PIN environment variable, so it should continue to do so. This would then be code in p11sak itself that first checks the environment variable and if not set prompt for the pin (using the new common pin functions). We will then need a force-pin-prompt option for p11sak, but only for that single tool.

holger-dengler commented:
The new version covers:
- rename of env variables
- removal of struct pinbuf (replaced with char **)
- usage of OPENSSL_cleanse() for pin cleanup
- fixes for the compare in pin_get_new()
- various fixes for review comments and typos

This is not the final version, some review comments are still open. But I think it is worth, to share this version.

holger-dengler commented:
> The more I think about that the more I come to the conclusion that the whole environment stuff in the pin handling code causes more trouble than it solves. Adding a force-pin-prompt option to all the tools is a nightmare.
> 
> Why not drop all the environment code from the pin handling code, and have the user use `echo $ENVVAR | tool` or `tool --pin $ENVVAR` if one wants to have the pin in an environment variable. No extra code needed for that.
> 
> The only exception would be p11sak. It is documented that it uses the PKCS11_USER_PIN environment variable, so it should continue to do so. This would then be code in p11sak itself that first checks the environment variable and if not set prompt for the pin (using the new common pin functions). We will then need a force-pin-prompt option for p11sak, but only for that single tool.

Fine with me. But again: There is no regression, also not for p11sak. No one can **ever** have used the environment variables there, because the implementation was broken! I had to look into the code, to figure out, how to trigger it. The env variable was only used, if you entered an empty pin on prompt. This was the initial trigger for this PR...

ifranzki commented on usr/lib/common/pin_get.c:
```diff
@@ -0,0 +1,114 @@
+    fflush(stdout);
+    echo(false);
+
+    n = getline(buf, &s, stdin);
```
I guess you need to take care about the situation that *buf is non-NULL on entry of the pin_prompt funmction.
getline accepts a non-null lineptr argument (aka buff), but then expects *n (aka s) to comtain the size if the buffer in *lineptr.

So either make sure that *buf can not be non-NULL when pin_prompt is called, or setup s like `s = strlen(*buf) + 1` when *buf is non NULL. Alternatively, just free *buf if it is not NULL, and let getline allocate a new one.

holger-dengler replied:
Since we're no longer using a struct, *buf != NULL is a coding error. So I would prefer to return NULL in these cases.

ifranzki commented:
If we keep the environment variable support, we MUST HAVE an option (e.g --force-pin-prompt) to force it to prompt for the pin, even if the environment variable is set. We would need this for all tools that use the new pin handling functions. 

Since adding this option to all the tools is quite an effort, I still think we should support the environment variable ONLY for p11sak (and add --force-pin-prompt for p11sak only).

holger-dengler commented:
I will remove the env support for all tool except p11sak. For p11sak, we have 2 options: a) introduce the --force-pin-prompt option or b) lookup the env variable only on empty pin-prompt entry (as it was before this PR).

This will also make all man page changes (excpt for p11sak) obsolete.

ifranzki commented:
> For p11sak, we have 2 options: a) introduce the --force-pin-prompt option or b) lookup the env variable only on empty pin-prompt entry

Option a please.

ifranzki commented:
> This will also make all man page changes (excpt for p11sak) obsolete.

It will additionally make the changes in the expect scripts obsolete (your very last commit).

holger-dengler commented:
Changes since last version:
- remove environment support for PKCS11_USER_PIN and PKCS11_SO_PIN for all tools, except for p1sak.
- restructure commits (cleanups first, new functionality second).
- drop all obsolete patches (e.g. changes to man-pages and expect-scripts).

ifranzki commented:
There are some failures in the p11sak tests in Travis. Please check.

ifranzki commented on usr/sbin/p11sak/p11sak.c:
```diff
@@ -25,7 +25,7 @@
 #include <pwd.h>
 
-#include <termios.h>
+#include <pin_prompt.h>
```
shouldn't this be `#include "pin_prompt.h"` since its a local header file?

holger-dengler replied:
No. The CFLAGS for p11sak are extended with `-I${srcdir}/usr/lib/common` and therefore the .h files in this directory can be included over the search path. This is the same in all other tools.

ifranzki replied:
I know that it works with <...>, otherwise we would see compile errors, but I would still prefer to include local headers using "..." and not <....>. Kind of a code style thing......

holger-dengler replied:
ok, But why are the CFLAGS extended then? Anyhow, I will change it here and in all other places.

ifranzki replied:
Without the extended CFLAGS it would not find it either way. 
The difference between <...> and "..." is pretty much only that "..." first looks in the directory of the current C-file, and then in the include path, whereas <...> ONLY looks in the include path. 
See https://gcc.gnu.org/onlinedocs/cpp/Search-Path.html
Since pin_prompt.h is not in the same directory as p11sak.c, it won't find it without 
`-I${srcdir}/usr/lib/common`

holger-dengler replied:
ok

ifranzki commented on testcases/unit/pintest.c:
```diff
@@ -0,0 +1,41 @@
+#include <stdlib.h>
+#include <unistd.h>
+
+#include <pin_prompt.h>
```
shouldn't this be `#include "pin_prompt.h"` since its a local header file?

ifranzki commented on usr/lib/common/pin_prompt.c:
```diff
@@ -0,0 +1,96 @@
+#include <sys/mman.h>
+#include <openssl/crypto.h>
+
+#include <pin_prompt.h>
```
shouldn't this be `#include "pin_prompt.h"` since its a local header file?

ifranzki commented on usr/sbin/pkcscca/pkcscca.c:
```diff
@@ -42,6 +41,7 @@
 
 #define OCK_TOOL
 #include "pkcs_utils.h"
+#include <pin_prompt.h>
```
shouldn't this be `#include "pin_prompt.h"` since its a local header file?

ifranzki commented on usr/sbin/pkcsconf/pkcsconf.c:
```diff
@@ -32,15 +31,16 @@
 #include "defs.h"
 #include "mechtable.h"
 #include "uri.h"
+#include <pin_prompt.h>
```
shouldn't this be `#include "pin_prompt.h"` since its a local header file?

ifranzki commented on usr/sbin/pkcsicsf/pkcsicsf.c:
```diff
@@ -32,6 +31,7 @@
 #include "defs.h"
 #include "cfgparser.h"
 #include "configuration.h"
+#include <pin_prompt.h>
```
shouldn't this be `#include "pin_prompt.h"` since its a local header file?

ifranzki commented on usr/sbin/pkcsep11_migrate/pkcsep11_migrate.c:
```diff
@@ -21,13 +21,13 @@
 #include "../../include/pkcs11types.h"
 #include "../../lib/common/p11util.h"
 #include "../../lib/ep11_stdll/ep11_func.h"
+#include <pin_prompt.h>
```
shouldn't this be `#include "pin_prompt.h"` since its a local header file?

ifranzki commented on usr/sbin/pkcsconf/pkcsconf.c:
```diff
@@ -260,136 +230,79 @@ int main(int argc, char *argv[])
+
+        if (!pin_new)
+            pin_new = pin_prompt_new(&buf_new,
+                                     "Enter new SO PIN: ",
```
Chang to `"Enter the new SO PIN: "` and `"Re-enter the new SO PIN: "`

ifranzki replied:
also check the other place to output the exact same message string as before in all cases.

ifranzki replied:
... also for the other tools. Changing the message text may cause regressions in existing scripts, also for distro partners, users, etc.

holger-dengler commented on usr/lib/common/pin_prompt.c:
```diff
@@ -0,0 +1,96 @@
+    ssize_t n;
+    size_t s;
+
+    if (!buf && *buf)
```
Ok, will fix it.

holger-dengler commented:
changes since last version:
- reuse pin prompt strings
- fixes for review comments

holger-dengler commented:
changes since lat version:
- rebased to current master

ifranzki reviewed (approved):
looks good now. Thanks a lot!

jschmidb reviewed (approved):
lgtm
